File: Streamy/Streamy/ViewController.m
//
//  ViewController.m
//  Streamy
//
//  Created by Arthur Myronenko on 3/3/16.
//  Copyright © 2016 Arthur Myronenko. All rights reserved.
//

#import "ViewController.h"
#import "AMServer.h"
#import "StreamManager.h"

@interface ViewController () <StreamManagerDelegate>

@property (weak) IBOutlet NSImageView *imageView;

@property (strong, nonatomic) AMServer *server;
@property (strong, nonatomic) StreamManager *streamManager;
@property (strong, nonatomic) NSTimer *timer;

@end

@implementation ViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    
    self.server = [[AMServer alloc] init];
    self.streamManager = [[StreamManager alloc] init];
}

- (IBAction)start:(NSButton *)sender {

    [self.streamManager startCaptureWithDelegate:self];
    [self.server connectToHost:@"localhost" onPort:8001];
//    self.timer = [NSTimer scheduledTimerWithTimeInterval:0.3 target:self selector:@selector(sendMessage) userInfo:nil repeats:YES];
}

- (IBAction)stop:(NSButton *)sender {
    [self.server disconnect];
    [self.streamManager stop];
    [self.timer invalidate];
}

- (void)sendMessage {
    [self.server sendImage:self.imageView.image];
    [self.server sendMessage:@"end_of_frame"];
}

#pragma mark - StreamManagerDelegate

- (void)streamManager:(StreamManager *)manager didCaptureFrame:(NSImage *)image {
    self.imageView.image = image;
    [self.server sendImage:self.imageView.image];
    [self.server sendMessage:@"end_of_frame"];
}

@end

File: Streamy/Streamy/ViewController.swift
//
//  ViewController.swift
//  Streamy
//
//  Created by Arthur Myronenko on 4/17/16.
//  Copyright © 2016 Arthur Myronenko. All rights reserved.
//

import Cocoa

class ViewController: NSViewController {

    @IBOutlet weak var imageView: NSImageView!
    
    private let server = StreamyServer()
    private let streamManager = StreamManager()
    private var started = false

    @IBAction func toggle(sender: NSButton) {
        if started {
            sender.title = "Start"
            stop()
        } else {
            sender.title = "Stop"
            start()
        }
    }
    
    private func start() {
        started = false
        streamManager.startCaptureWithDelegate(self)
        server.connectToHost("localhost", port: 8001)
    }
    
    private func stop() {
        started = false
        streamManager.stop()
        server.disconnect()
    }
}

extension ViewController: StreamManagerDelegate {
    func streamManager(streamManager: StreamManager, didCaptureFrame frame: NSImage) {
        imageView.image = frame
        server.sendImage(frame)
        server.sendMessage("end_of_frame")
    }
}